gluten-core/src/main/scala/io/glutenproject/extension/ExpandFallbackPolicy.scala

```diff
@@ -67,78 +67,93 @@ import org.apache.spark.sql.execution.exchange.{Exchange, ReusedExchangeExec}
         // support columnar, so the output columnar is always false in AQE postStageCreationRules
         case ColumnarToRowExec(s: Exchange) if isAdaptiveContext =>
-          countFallback(s)
+          countFallbackInternal(s)
         case u: UnaryExecNode
             if !u.isInstanceOf[GlutenPlan] && InMemoryTableScanHelper.isGlutenTableCache(u.child) =>
           // Vanilla Spark plan will call `InMemoryTableScanExec.convertCachedBatchToInternalRow`
           // which is a kind of `ColumnarToRowExec`.
           fallbacks = fallbacks + 1
-          countFallback(u.child)
+          countFallbackInternal(u.child)
         case ColumnarToRowExec(p: GlutenPlan) =>
           logDebug(s"Find a columnar to row for gluten plan:\n$p")
           fallbacks = fallbacks + 1
-          countFallback(p)
+          countFallbackInternal(p)
         case leafPlan: LeafExecNode if InMemoryTableScanHelper.isGlutenTableCache(leafPlan) =>
         case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
           // Possible fallback for leaf node.
           fallbacks = fallbacks + 1
-        case p => p.children.foreach(countFallback)
+        case p => p.children.foreach(countFallbackInternal)
       }
     }
-    if (!ignoreOneColumnarToRow(plan)) {
-      countFallback(plan)
-    }
+    countFallbackInternal(plan)
     fallbacks
   }
 
+  /**
+   * When making a stage fall back, it's possible that we need a ColumnarToRow to adapt to last
+   * stage's columnar output. So we need to evaluate the cost, i.e., the number of required
+   * ColumnarToRow between entirely fallback stage and last stage(s). Thus, we can avoid possible
+   * performance degradation caused by fallback policy.
+   *
+   * spotless:off
+   *
+   * Spark plan before applying fallback policy:
+   *
+   *        ColumnarExchange
+   *  ----------- | --------------- last stage
+   *    HashAggregateTransformer
+   *              |
+   *        ColumnarToRow
+   *              |
+   *           Project
+   *
+   * To illustrate the effect if cost is not taken into account, here is spark plan
+   * after applying whole stage fallback policy (threshold = 1):
+   *
+   *        ColumnarExchange
+   *  -----------  | --------------- last stage
+   *         ColumnarToRow
+   *               |
+   *         HashAggregate
+   *               |
+   *            Project
+   *
+   *  So by considering the cost, the fallback policy will not be applied.
+   *
+   * spotless:on
+   */
+  private def countStageFallbackCost(plan: SparkPlan): Int = {
+    var stageFallbackCost = 0
+
+    /**
+     * Find a Gluten plan whose child is QueryStageExec. Then, increase stageFallbackCost if the
+     * last query stage's output is columnar.
+     */
+    def countStageFallbackCostInternal(plan: SparkPlan): Unit = {
+      plan match {
+        case p: GlutenPlan if p.children.find(_.isInstanceOf[QueryStageExec]).isDefined =>
```

Comment:
thank you @PHILO-HE. I think we should only use this cost if all leaf node are QueryStageExec with columnar output, and we should also consider table cache. For example, if the fallback threshold is 1

before this pr
```
ColumnarShuffle       
      |                 
  QueryStage       Shuffle
      |               |
 ColumnarToRow    QueryStage
      \             /
            Join
```

after this pr
```
                   Shuffle
                      | 
                  QueryStage  
                      |
ColumnarShuffle  RowToColumnar
   |                  |
 QueryStage       QueryStage
    \               /
     JoinTransformer
           |
      ColumnarToRow   
```

Reply:
Thanks for your comment! Yes, it's possible that one last stage has row output. As RowToColumnar can be removed if current stage falls back, I think we can reduce the cost by 1 for such case to reflect the benefit of stage fallback. Just updated.

> we should also consider table cache.

Yes, just updated.